lib/chef/telemetry/base.rb

```diff
@@ -0,0 +1,120 @@
+        }
+      end
+
+      def run_starting(_opts = {})
+        Chef::Log.debug "Initiating telemetry for Chef"
+      end
+
+      def run_ending(opts)
+        payload = create_wrapper
+
+        # To not load ohai information once loaded
+        unless ohai
+          @ohai = Ohai::System.new
+          # Load plugins to gather system data
+          @ohai.all_plugins(%w{ os hostname platform dmi kernel})
+        end
+
+        payload[:platform] = ohai[:platform]
+
+        payload[:jobs] = [{
+                            type: JOB_TYPE,
+                            # Target platform info
+                            environment: {
+                              host: ohai[:hostname],
+                              os: ohai[:os],
+                              version: ohai[:platform_version],
+                              architecture: ohai[:kernel][:machine],
+                              id: (ohai[:dmi][:system] && ohai[:dmi][:system][:uuid]) || "",
+                            },
+                            runtime: Chef::VERSION,
+                            content: [],
+                            steps: [],
+                          }]
+
+        if opts[:run_context]
+          opts[:run_context].cookbook_collection.each do |_, value|
+            metadata = value.metadata
+            payload[:jobs][0][:content] << {
+              name: obscure(metadata&.name) || "",
+              version: metadata&.version || "",
+              maintainer: metadata&.maintainer || "",
+              type: "cookbook",
+            }
+          end
+          all_resources = opts[:run_context].resource_collection&.all_resources
+          if all_resources
+            all_resources.each do |resource|
+              payload[:jobs][0][:steps] << {
+                name: resource.recipe_name,
+                resources: [],
+              }
+
+              payload[:jobs][0][:steps].last[:resources] << {
+                type: "chef-resource",
+                name: resource.resource_name.to_s,
+              }
+            end
+          end
+        end
+        Chef::Log.debug "Final data for telemetry upload -> #{payload}"
+        Chef::Log.debug "Finishing telemetry for Chef"
+        # Return payload object for testing
+        payload
+      end
+
+      # TBD Should we implement distribution name based on below usage?
+      def determine_distribution_name
+        run_context = Chef::Telemetry::RunContextProbe.guess_run_context
+        case run_context
+        when "chef-zero"
+          ChefUtils::Dist::Zero::EXEC
+        when "chef-apply"
+          ChefUtils::Dist::Apply::EXEC
+        when "chef-solo"
+          ChefUtils::Dist::Solo::EXEC
+        else
+          ChefUtils::Dist::Infra::EXEC
+        end
+      end
+
+      # Hash text if non-nil
+      def obscure(cleartext)
+        return nil if cleartext.nil?
+        return nil if cleartext.empty?
+
+        Digest::SHA2.new(256).hexdigest(cleartext)
```

Comment:
Do you think implementing salting with this might be helpful?

```
salt = SecureRandom.hex(16)  # Generate a random 16-byte salt
hash = Digest::SHA2.new(256).hexdigest(salt + cleartext)
```

Reply:
@dafyddcrosby - the entire intent on capturing the SHA of the without salt was to actually have that rainbow table.

We are looking to understand, what cookbooks are being used from public supermarket. Ideally with this information we can track which cookbook (and version). That are used. That will help us understand what are the most popular cookbooks (download data is not enough for that), which cookbooks being used but are no longer maintained, and as the usage data is tried to a license we could explore the possibility of notify people of cookbook updates/vulnerabilities (inspired from depend-a-bot). 

The thought process here was cookbook names could contain sensitive information, so transmitting it without a hash was not acceptable, however adding in the salt would make this data meaningless, unless the salt was part of your license (which would also downgrade the usefulness of the salt). 

We are not interested in collecting information on private cookbooks or user-specific content. I am open to any suggestions on alternative ways to collect this information?

Reply:
> however adding in the salt would make this data meaningless, unless the salt was part of your license (which would also downgrade the usefulness of the salt).

With no salt, one could just grab a rainbow table off https://crackstation.net/ and call it a day. Whereas if you used eg the serial off the trial key as a salt, while you could still quickly generate one-off rainbow tables of cookbooks out of the known cookbook names in supermarket, it increases the cost of finding *unknown* cookbook names.

Conceptually, the client sends a set of hashes along with the serial. Server-side, you have something like this pseudocode:
```
def find_supermarket_cookbooks(serial, set)
  table = SUPERMARKET_COOKBOOK_NAMES.to_h {|cb| [Digest::SHA2.new(256).hexdigest(serial + cb), cb] }
  set.map! {|cb_hash| table[cb_hash]}
  set.compact!
  set
end
```

It's not a perfect solution, but gets you the information you want re: supermarket cookbooks, and is still largely cacheable so that you don't need to recompute hashes on the backend after the first upload from a license.

Reply:
@dafyddcrosby  - one question and one comment

Question: what is the concern about someone being able to reverse some of the cookbook names?

Comment: If I am understanding your suggestion correctly (salt by license) that would have to result in us having one rainbow table per license issued. Am I understanding this correctly?